spec/formatting.md

```diff
@@ -505,14 +505,17 @@ _Formatting_ is a mostly implementation-defined process,
 as it depends on the implementation's shape for resolved values
 and the result type of the formatting.
 
-Formatting errors MAY be emitted during _formatting_,
-as formatting is not necessarily defined on every resolved value.
-A formatter MAY provide a value to use in such a case instead of a _fallback value_.
+Formatting is not necessarily defined on every resolved value.
+An implementation SHOULD emit formatting errors during _formatting_,
+but MAY provide a value to use in such a case instead of a _falback value_.
 
-_Formatting_ MAY produce formatted messages with the following data types,
-as well as any others:
+An implementation MAY use the value of an _expression_'s _operand_
+to choose a _formatting function_ and default _options_ if no
+_annotation_ is supplied by the _expression_.
```

Comment:
I think this belongs in the Variable Resolution section, not here. As I understand it, we want to make it possible for something like `{$num}` and `{$num :number}` to format the same, if `$num` is a number.

To achieve that, it's much easier to say that an implementation may apply a default annotation to a variable reference that refers to an external variable, depending on its type and value.

If we add this here, then a conforming implementation would need to include in its resolved-value shape information about each value's original expression shape, which is otherwise unnecessary. That's an unnecessary restriction on implementations.

Reply:
I thought about that initially. One challenge is the need to do a bunch of changes if we do that. Currently the parent "expression resolution" section says:

> Depending on the presence or absence of an operand and a function, private-use, or reserved annotation, one of the following is used to resolve the value of the expression:
> *  If the expression contains no annotation, its resolved value is determined by literal resolution or variable resolution, depending on the shape of the operand.
> *  Else, if the expression has a function annotation, its resolved value is defined by function resolution.
> * _... more else clauses dealing with private/reserved..._

I think that we can't ignore literals. It's one thing if a literal is just a blob of text, but what about:

```
let $when = {|2023-07-15T00:00:00Z|}
let $howMany = {4200}
{You bought {$howMany} on {$when}}
```

You might say that this is variable resolution, but then what about:

```
{You bought {4200} on {2023-07-16}}
```

If we want to allow (but not force) implementations to infer a type and thus a formatter, we would need to recast literal and variable resolution to pass through function resolution. I suspect this might be the right thing to do anyway (the ultimate default formatter is the one that places stringified literals or stringified variables) rather than having them be separate paths as they are now.

Reply:
I think that a literal without a function is just the literal, there is no good way to infer a function for it.
At least not "in the engine" (in MF2 code itself), while still remaining function agnostic.

Reply:
@mihnita Fair enough. But remember that the keys of a selector are literals:

```
when 0 {This is the =0 message}
```

If a literal is always and *only* a character sequence, how can I assign a number to a variable? A date?

Of course there are sticky issues. What does this print?

```
{That is {NaN :number}}  <- CLDR has data for formatting this
```

Reply:
I think we should require an explicit function for formatting literals; by default they should always end up as strings.

> If a literal is always and _only_ a character sequence, how can I assign a number to a variable? A date?

By having a `:number` or `:datetime` function that can accept string input. For example `{42 :number}`.

> Of course there are sticky issues. What does this print?
> 
> ```
> {That is {NaN :number}}  <- CLDR has data for formatting this
> ```

This is implementation-dependent behaviour, unless we define the supported input values of `:number` rather explicitly. With the JS proposal in its current shape, you'd end up with `That is {|NaN|}` because it only accepts JSON numerical values for string input.

Reply:
We can _allow_ implementations to "cast" literals that they recognize. For many implementations and users this will seem natural and many implementations will probably do mainly the right thing here. The alternative, as you point out @eemeli, would be to require non-literals to use an annotation.

This is, to some degree, a loss of functionality for strongly typed languages vs. MF1, where a pattern like `You have {0}` will format a `Number` or a `Date` (or its `Temporal` or `Calendar` friends) as those types when they are passed and not just call `toString()`.

The prose I'm proposing would allow (MAY) this but not require it. Do you object to that?

Reply:
At the moment, we say this [in the spec](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#literal-resolution):
> The resolution of a _text_ or _literal_ token MUST always succeed.

If we allow for casting literals to non-string types, we'll need to remove that assertion, and we will no longer be able to say with certainty e.g. that this message will always format to the same string in all implementations:

```
{Enter code {12345} to continue}
```

That could come out as e.g. any of the following, even in just "English":
- Enter code 12 345 to continue
- Enter code 12,345 to continue
- Enter code 12.345 to continue

> This is, to some degree, a loss of functionality for strongly typed languages vs. MF1, where a pattern like `You have {0}` will format a `Number` or a `Date` (or its `Temporal` or `Calendar` friends) as those types when they are passed and not just call `toString()`.

This is not the case. What we do with unannotated literals does not affect what we do with unannotated variable references. If a variable resolves to a number, we can use number formatting for it. Same if it's a Date, or any other value type that the implementation has defined handling for.

> The prose I'm proposing would allow (MAY) this but not require it. Do you object to that?

Yes, because I think that allowing such value casts adds unnecessary uncertainty without a clear and significant benefit.

Reply:
> @mihnita Fair enough. But remember that the keys of a selector are literals:

But they have a function that looks at them and interprets them.
That is the matched function.

In the `when 0 {This is the =0 message}` it is the `plural` function.
As the matcher functions take the key(s), the input, and tell us if they match or not.
So it is still a function doing the interpretation and the casting.
And if one implements a `choice` matcher, it would be that function looking at the key in `when |<10| {...}`

Reply:
> let $x = 12345

`$x` contains a literal at this point.

> let $x = $countFromArgument

`$x` might contain a number here. If the programming languages have some kind of type system.

Note that even MF1 for C has type information here, as the real values to be formatted are wrapped in a 
[`Formattable`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1Formattable.html)

Reply:
> > If we allow for casting literals to non-string types, we'll need to remove that assertion
> 
> I disagree, I think. The _resolution_ of a text or literal would always succeed, but the formatting might be different. These are separate operations.

We're getting into semantics, but you are indeed right if we consider the string-without-pattern-matching-applied as "resolved". In other words, we could leave the assertion in, but we would need to add in a much more restrictive definition of "resolved value" than we currently have, because that could not include the resolution of a string `'12345'` as a numerical value.

> > we will no longer be able to say with certainty e.g. that this message will always format to the same string in all implementations
> 
> It is already the case that the same message might not format to the same string in all implementations--because, for example, they are running different CLDR versions or the locale resolves slightly differently on a given platform.

In general yes, but my point was about the specific string including an explicit literal. It does currently format to the same string in all possible implementations.

> The example you provided with `{12345}` is a good one, since it highlights the problem of untyped or weakly typed values, which this proposal brings to the fore.

In my understanding, the value represented in `{12345}` is always a string, and only ever a string. Its value could be parsed as a number, but it's still a string.

> > > This is, to some degree, a loss of functionality for strongly typed languages vs. MF1
> >
> > This is not the case
>
> No, I'm pretty sure it is the case. MF1 lets you write simple patterns that don't declare a formatter and uses reflection to assign the formatter at runtime, e.g. `{0}` formats a `Number`, `Date`, or `String` differently and using the appropriate formatter. Whether that capability is important is what we're discussing 😉.

Just as a sanity check, I'm presuming that the `{0}` you're using is an MF1 reference to an external variable, yes? If so, that's completely separate from this discussion on literal values that are wholly and completely defined in the message syntax.

> Wouldn't you feel a bit of surprise if you didn't get the same behavior when you did a local assignment?

Not if we make it clear that the value of an unannotated literal is always a string.

> On the one hand, I do think that requiring the annotation is super helpful (particularly to translators, who cannot see what type is being passed), but that suggests that we require an annotation all the time.

The we should also define a `:string` function.

> The question of clear and significant benefit is whether developers can get "do what I mean" behavior. We should seek group consensus here.

Agreed.